### PR TITLE
mw-python-tools: major improvements

### PR DESCRIPTION
**Pull request by RhinosF1**
Deploy mediawiki changes 
- new WikiCommand class
- Checks non-mediawiki commands don't run when mediawiki doesn't exist
- New l10n command method
- introduces async command running
- various style changes
- exitcodes is now an array passed throughout
- remote_sync_file returns list[int] not int
- run now returns the exitcode but doesn't exit
- More run() variables type checked
- run() uses async commands where possible and WikiCommand for mediawiki commands
- prep() has been split out of run()
- new deployment map introduced!
- --servers is now validated
- get_server_list() now takes the list of available servers as parameter 1 from Environment['servers'] rather than Environment.
- unknown arguments are now discarded.
- replaced remote_sync_file with _prep and moved actual running to run() (commands now in map)
- split getting debug url out to ease an extra few tests

Testing:

I encourage deploys on test101 in case I've missed anything but the best way to see what it will do is by using the deployment maps.

Please feel free to send a list of deploys you'd like maps verifying for.

The maps are there to give an insight into what it will do when you run it.

### CODE REVIEW

**Comment by RhinosF1** [comment]
DO NOT MERGE

**Comment by RhinosF1** [comment]
@JohnFLewis: Please re-assign to @Universal-Omega so they can code review when back

**Comment by RhinosF1** [comment]
Would an actual --no-op parameter be useful?

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 1]
```diff
@@ -167,134 +236,151 @@ def _construct_git_pull(repo: str, submodules: bool = False, branch: Optional[st
...
-    if HOSTNAME in servers:
-        del loginfo['servers']
-        text = f'starting deploy of "{str(loginfo)}" to {synced}'
-        if not args.nolog:
+    deploymentmap['loginfo'] = loginfo
+    deploymentmap['commands']['stage'] = _get_git_commands(args.world, args.pull, args.branch)
+    for option in options:  # configure rsync & custom data for repos
+        if options[option]:
+            if option == 'world':  # install steps for w
+                deploymentmap['commands']['rebuild'].append(WikiCommand('MW_INSTALL_PATH=/srv/mediawiki-staging/w php /srv/mediawiki-staging/w/extensions/MirahezeMagic/maintenance/rebuildVersionCache.php --save-gitinfo --conf=/srv/mediawiki-staging/config/LocalSettings.php', envinfo['wikidbname']))
```
Comment (Universal-Omega) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:307]: Would this not essentually be `sudo -u www-data php MW_INSTALL_PATH=... php...`?

**Review by Universal-Omega — CHANGES_REQUESTED** [review]

File: modules/mediawiki/files/bin/deploy-mediawiki.py (hunk 1 above)
Comment (RhinosF1) [reply modules/mediawiki/files/bin/deploy-mediawiki.py:307]: Yeah, I'll fix that

File: modules/monitoring/files/check_reverse_dns.py [hunk 2]
```diff
@@ -37,29 +39,29 @@ def get_args():
...
-        description='Check reverse DNS entry for hostname'
+        description='Check reverse DNS entry for hostname',
     )
     parser.add_argument(
         '-H',
         '--hostname',
         required=True,
         help='hostname to check',
-        dest='hostname'
+        dest='hostnam',
```
Comment (Universal-Omega) [inline modules/monitoring/files/check_reverse_dns.py:49]:
```suggestion
        dest='hostname',
```

File: modules/monitoring/files/bot/ircecho.py [hunk 3]
```diff
@@ -12,29 +12,30 @@
...
 def beautify_message(m):
-    '''Clean up formatting of alert messages.'''
+    """Clean up formatting of alert messages."""
     m = m.strip()                           # Strip trailing whitespace
     m = re.sub(r'(\w+): \1:\b', r'\1', m)   # Dedupe severity
     m = re.sub(r' {2,}', ' ', m)            # Collapse whitespace
     m = m.replace(': -', ':')               # Combine separators
-    m = m.strip(':-')                       # Strip trailing separators
-    return m
+    return m.strip(':-')                       # Strip trailing separators
```
Comment (Universal-Omega) [inline modules/monitoring/files/bot/ircecho.py:38]:
```suggestion
    return m.strip(':-')                    # Strip trailing separators
```

**Review by Universal-Omega — CHANGES_REQUESTED** [review]

**Comment by RhinosF1** [comment]
What no-op means
```
{'branch': '', // the branch if --branch given - NEEDED?
 'commands': {'postinstall': [], 
              'rebuild': [WikiCommand(command='MW_INSTALL_PATH=/srv/mediawiki-staging/w /srv/mediawiki-staging/w/extensions/MirahezeMagic/maintenance/rebuildVersionCache.php --save-gitinfo --conf=/srv/mediawiki-staging/config/LocalSettings.php', wiki='testwiki'),
                          WikiCommand(command='/srv/mediawiki/w/extensions/CreateWiki/maintenance/rebuildExtensionListCache.php', wiki='testwiki')],
              'rsync': ['sudo -u www-data rsync --update -r --delete '
                        '--exclude=".*" /srv/mediawiki-staging/w/* '
                        '/srv/mediawiki/w/'],
              'stage': ['sudo -u www-data git -C /srv/mediawiki-staging/w/ '
                        'pull --recurse-submodules --quiet']
}, // above are the command to run on mwtask111
 'debugurl': 'publictestwiki.com', // wiki to use for canary checks
 'doworld': True, // if --world is passed so composer can be ran
 'force': False, // if --force was passed
 'ignoretime': False, // if --ignore-time was passed - NEEDED?
 'loginfo': {'extensionlist': True,
             'noop': True,
             'servers': 'mw101',
             'world': True}, // info to format the !log message to IRC
 'nolog': False, // whether to log to shell or IRC
 'port': None, // port (implies http(s):// for canary checks)
 'remote': {'commands': ['sudo -u www-data rsync --update -r --delete -e "ssh '
                         '-i /srv/mediawiki-staging/deploykey" '
                         '/srv/mediawiki/cache/gitinfo/ '
                         'www-data@mw101.miraheze.org:/srv/mediawiki/cache/gitinfo/',
                         'sudo -u www-data rsync --update -r --delete -e "ssh '
                         '-i /srv/mediawiki-staging/deploykey" '
                         '/srv/mediawiki/w/ '
                         'www-data@mw101.miraheze.org:/srv/mediawiki/w/',
                         'sudo -u www-data rsync --update -e "ssh -i '
                         '/srv/mediawiki-staging/deploykey" '
                         '/srv/mediawiki/cache/extension-list.json '
                         'www-data@mw101.miraheze.org:/srv/mediawiki/cache/extension-list.json'],
            'files': ['/srv/mediawiki/cache/extension-list.json'],
            'paths': ['/srv/mediawiki/cache/gitinfo/', '/srv/mediawiki/w/']},
// the commands, files and paths to sync too
 'servers': ['mw101']} // the servers synced too - NEEDED?

```

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 4]
```diff
@@ -167,138 +241,156 @@ def _construct_git_pull(repo: str, submodules: bool = False, branch: Optional[st
...
-        rsyncpaths.append('/srv/mediawiki/cache/l10n/')
-
-    for path in rsyncpaths:
-        exitcodes.append(remote_sync_file(time=args.ignoretime, serverlist=servers, path=path, force=args.force, envinfo=envinfo, nolog=args.nolog))
-    for file in rsyncfiles:
-        exitcodes.append(remote_sync_file(time=args.ignoretime, serverlist=servers, path=file, recursive=False, force=args.force, envinfo=envinfo, nolog=args.nolog))
-
-    fintext = f'finished deploy of "{str(loginfo)}" to {synced}'
+            if option == 'world':  # install steps for w
+                deploymentmap['commands']['rebuild'].append(WikiCommand('MW_INSTALL_PATH=/srv/mediawiki-staging/w /srv/mediawiki-staging/w/extensions/MirahezeMagic/maintenance/rebuildVersionCache.php --save-gitinfo --conf=/srv/mediawiki-staging/config/LocalSettings.php', envinfo['wikidbname']))
```
Comment (Universal-Omega) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:312]: Would this not be `sudo -u www-data php MW_INSTALL_PATH=/srv/mediawiki-staging/w ...` still? Whereas it should be `sudo -u www-data MW_INSTALL_PATH=/srv/mediawiki-staging/w php ...` I think? (The env before php)

**Review by Universal-Omega — CHANGES_REQUESTED** [review]

File: modules/mediawiki/files/bin/deploy-mediawiki.py (hunk 4 above)
Comment (RhinosF1) [reply modules/mediawiki/files/bin/deploy-mediawiki.py:312]: Yes, that's how wiki command will output it. Can you test to see if that breaks stuff? I can add support for injecting the env variable if not.

**Comment by JohnFLewis** [comment]
No recent progress and merge conflicts need to be resolved.

Closing - please re-open with progress.

**Pull request closed without merging**